### ISSUE TITLE
G8OCREngineModeTesseractCubeCombined and G8OCREngineModeCubeOnly issues for rus locale

### ISSUE THREAD

**Issue by ws233** [gali8/Tesseract-OCR-iOS#140]
1. Setup tesseract with rus locale.
Set G8OCREngineModeTesseractCubeCombined. Try to recognize.
Crash: init_cube_objects(true, &tessdata_manager):Error:Assert failed:in file tessedit.cpp, line 203

2. Setup tesseract with rus locale.
Set G8OCREngineModeCubeOnly. Try to recognize.

The corresponding code snippet is in the it(@"Should set variables from dictionary and reinit correctly", ^{ unit test block from the #139 PR. Just unccomment the line 348

There is also a recognized blocks image file attached. It seems that blocks there are correct, but their confidences are wrong.

Does anybody knows, why tess doesn't work with rus locale in cube or combined mode?

**Comment by kevincon** [comment]
I was able to reproduce everything you described.

It's very odd, because if I execute the following from the command line (with Tesseract 3.03 installed and all the "rus" traineddata/cube files in my computer's tessdata folder):

```
tesseract image_sample.jpg stdout -l rus -c tessedit_ocr_engine_mode=1
```

I (correctly) get:

```
1234567890

```

I googled around and played with some of the test settings but I can't figure out what might be happening. I even modified Template Framework Project to use "rus" and the cube/cube-combined engine modes and it still did the same thing as in the tests.

I'd recommend reporting an issue to the upstream Tesseract OCR library, but since it seems to work using their binary, that suggests that it's an issue with our library...sorry I can't be of more help. :(

**Comment by ws233** [comment]
Thanks for this investigation, @kevincon.
Just a few questions to make it more clear for me.
rus traineddata from the host tesseract and from the iOS app bundle are exactly the same? Did you compare them? If you use brew to install tesseract, there could be different traineddata files.

**Comment by kevincon** [comment]
Yeah that's one of the things I tried. I copied the rus files from my computer's installation (which came from homebrew) to the iOS referenced folder and it didn't change the result. I didn't try it the other way around (copying iOS files to my computer), but I don't think that would have worked either.

**Comment by ws233** [comment]
ok. I see. I will try to search for a decision before submitting an issue to upstream Tes.

**Comment by ws233** [comment]
Also, I didnot find -c parameter parsing in tesseractmain.cpp.
Are you sure the command line above initiate Cube mode?
Could you give me a favour and try creating a config file with the variable above and then send the config file name in the command line instead of the -c parameter?

**Comment by ws233** [comment]
I've found a reason.
If you create a tess instance with G8OCREngineModeTesseractCubeCombined everything works fine and tests pass. So it seems that we do not do some clearing while changing engineMode.
And it seems that cube_cntxt_ should be cleared. Take a look on line 204 of the cube_controll.cpp. That seems to be a place of assertion. It's similar to the print log from my first post.

**Comment by ws233** [comment]
Uooops! Sorry for that. I've used eng locale. But for rus. it still doesn't work.
A few more logs with cube_debug_level=1
Cube ERROR (Failed to instantiate TesseractCubeCombiner
init_cube_objects(true, &tessdata_manager):Error:Assert failed:in file tessedit.cpp, line 203
So we can see that this function runs till the end and return false.

**Comment by kevincon** [comment]
Aha! You were right, the `-c` parameters aren't respected despite the man page saying they should be.

When I created a config file to set the engine mode to cube only, I got the same result we do in iOS:

```
1234669890

```

And when I set the engine mode to combined with cube, I get the same crash that we do on iOS:

```
init_cube_objects(true, &tessdata_manager):Error:Assert failed:in file tessedit.cpp, line 203
[1]    5562 abort      tesseract image_sample.jpg stdout -l rus rus-test
```

Therefore, I think you are definitely justified in submitting an issue to the upstream Tesseract library. :)

**Comment by ws233** [comment]
ok. I will. But I think they do not answer anymore.

**Comment by ws233** [comment]
Hm. I think that's the reason.
There are two files for eng locale: eng.tesseract_cube.nn and eng.cube.nn.
And only one for rus locale. rus.cube.nn.
It seems there is one file missing :(

**Comment by ws233** [comment]
I tried just to remove eng.tesseract_cube.nn to rus.tesseract_cube.nn and crach disappeared.
But now, G8OCREngineModeTesseractCubeCombined produces the same output as G8OCREngineModeCubeOnly.

**Comment by ws233** [comment]
I've created an issue for upstream tesseract here:
https://code.google.com/p/tesseract-ocr/issues/detail?id=1492&thanks=1492&ts=1435516671
Let's follow it.

**Comment by rmtheis** [comment]
Good catch. You should consider opening this issue on the new Github issues page for Tesseract: https://github.com/tesseract-ocr/tesseract/issues

**Comment by kevincon** [comment]
That's great that they're on GitHub now. @ws233 since you're the main stakeholder in this issue, can you recreate it there?

**Comment by ws233** [comment]
@kevincon, it seems that a lot of people tries Cube mode and gets craches with it. What do you think about deprecating it, since it's no longer supported as it was mentioned in upstream repo? I'm thinking about removing the corresponding property and corresponding parameter from all initializers since they are definitely useless.

**Comment by ws233** [comment]
Arabic works only in Cube mode. So it cannot be removed.

**Comment by Munderkind** [comment]
Arabic language normally work only in G8OCREngineModeCubeOnly. Murphy's Law in practice, I tried G8OCREngineModeTesseractOnly and G8OCREngineModeTesseractCubeCombined and I getting errors.

**Comment by kevincon** [comment]
@ws233 I agree that Cube is causing more trouble than its worth. But since I don't know when the upstream repo will produce another release that removes Cube, what if we just raise an exception/error if the user tries to call `recognize:` with `engineMode` set to something other than `G8OCREngineModeTesseractOnly` and with a `language` other than "eng"? My understanding is that Cube does work for English, so I don't think we should break that support.
